lightningd/pay.c

```diff
@@ -596,37 +596,36 @@ send_payment(struct lightningd *ld,
 	enum onion_type failcode;
 	size_t i, n_hops = tal_count(route);
 	struct hop_data *hop_data = tal_arr(tmpctx, struct hop_data, n_hops);
-	struct pubkey *path = tal_arr(tmpctx, struct pubkey, n_hops);
 	struct node_id *ids = tal_arr(tmpctx, struct node_id, n_hops);
 	struct wallet_payment *payment = NULL;
 	struct htlc_out *hout;
 	struct short_channel_id *channels;
 	struct routing_failure *fail;
 	struct channel *channel;
+	struct sphinx_path *path;
+	struct short_channel_id finalscid;
+	struct pubkey pubkey;
+	bool ret;
 
 	/* Expiry for HTLCs is absolute.  And add one to give some margin. */
 	base_expiry = get_block_height(ld->topology) + 1;
 
-	/* Extract IDs for each hop: create_onionpacket wants array of *keys*,
-	 * and wallet wants continuous array of node_ids */
-	for (i = 0; i < n_hops; i++) {
+	path = sphinx_path_new(tmpctx, rhash->u.u8);
+	/* Extract IDs for each hop: create_onionpacket wants array. */
+	for (i = 0; i < n_hops; i++)
 		ids[i] = route[i].nodeid;
-		/* JSON parsing checked these were valid, so Shouldn't Happen */
-		if (!pubkey_from_node_id(&path[i], &ids[i])) {
-			return command_fail(cmd, PAY_UNSPECIFIED_ERROR,
-					    "Invalid nodeid %s",
-					    type_to_string(tmpctx,
-							   struct node_id,
-							   &ids[i]));
-		}
-	}
```

Comment:
Removing this sanity check makes me uncomfortable.  But why don't we simply make `ids` a struct pubkey array directly, and do the `pubkey_from_node_id(&ids[i], &route[i].nodeid))` here in the loop and then fail immediately like this did?